lambda/matching.ml

```diff
@@ -2722,6 +2717,65 @@ let mk_failaction_pos partial seen ctx defs =
         | Some lam -> string_of_lam lam
         );
     (fail, [], jumps)
+  ) else (
+    (* Compute a specialized context for each failure pattern. *)
+    let fail_pats =
+      List.map (fun pat -> (pat, Context.lub pat ctx)) fail_pats in
+    (* Determine which exit is appropriate for which failure patterns. *)
+    let rec exit_failpats defs fail_pats acc =
+      match Default_environment.pop defs with
+      | Some ((idef, pss), rem) ->
+          (* Collect the patterns whose context matches the
+             matrix [pss] of the exit [idef]. *)
+          let now, later =
+            List.partition_map (fun ((p, p_ctx) as fail_pat) ->
+              if Context.matches p_ctx pss
+              then Either.Left p
+              else Either.Right fail_pat
+            ) fail_pats
+          in
+          let acc =
+            if now = [] then acc
```

Comment:
How would you feel about pattern-matching on the list instead of using an equality test ?
Or `List.is_empty`, since we have that now.

Reply:
It is always better to go in the direction of the reviewer (so I will use `List.is_empty` here), but I don't find your proposals clearer than the current code. If you are thinking of the performance impact, maybe we should consider optimizing polymorphic comparisons to constants?

Reply:
I'm not concerned about performance, but the previous code used pattern-matching, you have a pattern-matching that does the same thing 4 lines below, so I wondered why you chose to use the equality for this case only. (Some parts of the compiler re-export the `(=)` function with type `int -> int -> bool`, to prevent misuses of the polymorphic comparison, so I tend to be a bit defensive about it).

Reply:
The consistency with the other line below is a good point. I rewrote the code to use a pattern-matching.